debian/foo-yc20.1
.\"                                      Hey, EMACS: -*- nroff -*-
.TH FOO-YC20 1 "December  9, 2010"
.SH NAME
foo-yc20 \- YC-20 organ emulation
.SH SYNOPSIS
.B foo-yc20
.SH DESCRIPTION
The YC-20 is a divide-down combo organ designed in the late 60's. This
emulation faithfully copies the features, sounds and flaws of the original
organ.
.SH AUTHOR
foo-yc20 was written by Sampo Savolainen <v2@iki.fi>.
.PP
This manual page was written by Alessandro Ghedini <al3xbio@gmail.com>,
for the Debian project (and may be used by others).


debian/changelog
foo-yc20 (1.4.0~alxwk.git20240207-1) UNRELEASED; urgency=medium

  * Non-maintainer upload.
  * More historically accurate position for "Pitch" wheel

 -- alxwk <awjd15+gh@gmail.com>  Wed Feb  7 20:54:54 +05 2024

foo-yc20 (1.4.0~alxwk.git20240203-1) UNRELEASED; urgency=medium

  * Non-maintainer upload.
  * Local catch-up to upstream

 -- alxwk <awjd15+gh@gmail.com>  Sat Feb  3 20:42:33 +05 2024

foo-yc20 (1.3.0-7) unstable; urgency=medium

  * Team upload (Closes: #980807)

  [Olivier Humbert]
  * d/watch: update repo + version 4.
  * d/control: update homepage link.
  * d/copyright: http -> https.

  [Ondřej Nový]
  * d/compat: use debhelper-compat instead of debian/compat .
  * d/control: set Vcs-* to salsa.debian.org .
  * d/copyright: use https protocol in Format field.

  [Felipe Sateler]
  * d/control: change maintainer address to debian-multimedia@l.d.o

  [Pino Toscano]
  * Drop menu file, since foo-yc20 already provides a .desktop file.

 -- Alessandro Ghedini <ghedo@debian.org>  Sat, 01 Mar 2014 12:28:47 +0100

foo-yc20 (1.3.0-6) unstable; urgency=medium

  * Team upload
  * Remove myself from Uploaders
  * Fix MIDI controller number for the "Pitch" control.
    Thanks to Alex Volkov for the patch (Closes: #712114)
  * Bump Standards-Version to 3.9.5 (no changes needed)
  * Bump debhelper compat level to 9
  * Drop outdated DM-Upload-Allowed header

 -- Alessandro Ghedini <ghedo@debian.org>  Sat, 01 Mar 2014 12:28:47 +0100

foo-yc20 (1.3.0-5) unstable; urgency=low

  * Replace lv2core with lv2-dev on Build-Depends.

 -- Alessio Treglia <alessio@debian.org>  Mon, 14 May 2012 18:29:08 +0200

foo-yc20 (1.3.0-4) unstable; urgency=low

  [ Alessandro Ghedini ]
  * Email change: Alessandro Ghedini -> ghedo@debian.org

  [ Alessio Treglia ]
  * Temporarily disable DSP sources re-generation to prevent SIGSEGV
    at startup. See upstream bug report for more information:
    http://code.google.com/p/foo-yc20/issues/detail?id=5
  * Update debian/copyrignt to format 1.0.
  * Bump Standards.

 -- Alessio Treglia <alessio@debian.org>  Thu, 26 Apr 2012 11:06:28 +0200

foo-yc20 (1.3.0-3) unstable; urgency=low

  * Get rid of Intel optimizations and -m{tune,arch}=native
    flags (Closes: #649461).

 -- Alessio Treglia <alessio@debian.org>  Tue, 29 Nov 2011 08:33:38 +0100

foo-yc20 (1.3.0-2) unstable; urgency=low

  * Add 0003-include-faust-dir.patch (Closes: #642662)
  * Bump Standards-Version to 3.9.2 (no changes needed)

 -- Alessandro Ghedini <al3xbio@gmail.com>  Tue, 01 Nov 2011 17:43:19 +0100

foo-yc20 (1.3.0-1) unstable; urgency=low

  * New upstream release:
    - Refactor UI codebase.
  * Refresh 0002-makefile.patch
  * debian/rules:
    - Rebuild gen/*.cpp with faust (Closes: #613044).
    - Save & restore original files.
  * debian/control:
    - Build-Depends on faust.

 -- Alessio Treglia <alessio@debian.org>  Mon, 14 Feb 2011 01:02:55 +0100

foo-yc20 (1.2.1-1) unstable; urgency=low

  * Imported Upstream version 1.2.1
  * Fixes denormalization issues in the LV2 plugin

 -- Adrian Knoth <adi@drcomp.erfurt.thur.de>  Tue, 25 Jan 2011 14:20:13 +0100

foo-yc20 (1.2.0-1) unstable; urgency=low

  [ Adrian Knoth ]
  * Improve package description.

  [ Alessio Treglia ]
  * New upstream release.
  * Update debian/copyright.
  * debian/rules:
    - Set buildsystem to makefile.
    - Rely on dh_auto_build rather than calling make.
    - Append -Wl,--as-needed to LDFLAGS.
  * debian/control:
    - Add Provides: lv2-plugin
    - Replace libgtkmm-2.4-dev with libgtk2.0-dev, build-depend on
      lv2core.
  * debian/patches/0002-makefile.patch: Pass LDFLAGS to the linker.

 -- Alessio Treglia <alessio@debian.org>  Thu, 20 Jan 2011 21:44:18 +0100

foo-yc20 (1.1.0-1) unstable; urgency=low

  * Initial release. (Closes: #605785)

 -- Adrian Knoth <adi@drcomp.erfurt.thur.de>  Wed, 08 Dec 2010 23:20:12 +0100


src/foo-yc20.cpp
/*
    Foo-YC20 UI utility class

Copyright 2010-2011 Sampo Savolainen (v2@iki.fi). All rights reserved.
Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

   1. Redistributions of source code must retain the above copyright notice, 
      this list of conditions and the following disclaimer.

   2. Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation
      and/or other materials provided with the distribution.

   3. Neither the name Foo YC20, its author, nor the names of its contributors
      may be used to endorse or promote products derived from this software
      without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY SAMPO SAVOLAINEN ``AS IS'' AND ANY EXPRESS OR
IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
SHALL SAMPO SAVOLAINEN OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHERIN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
ADVISEDOF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include <iostream>
#include <fstream>
#include <cstring>
#include <cerrno>
#include <list>
#include <set>

#include <stdlib.h>


#include <foo-yc20.h>

void
YC20Processor::addButton(const char* label, FAUSTFLOAT* zone)
{
	bool isNote = true;
	int note;
	int octave = atoi(label+1);

	if (strlen(label) != 2) {
		isNote = false;
	}

	if (octave == 0 && errno == EINVAL) {
		isNote = false;
	}

	switch(label[0]) {
		case 'c': note =  0; break;
		case 'C': note =  1; break;
		case 'd': note =  2; break;
		case 'D': note =  3; break;
		case 'e': note =  4; break;
		case 'f': note =  5; break;
		case 'F': note =  6; break;
		case 'g': note =  7; break;
		case 'G': note =  8; break;
		case 'a': note =  9; break;
		case 'A': note = 10; break;
		case 'b': note = 11; break;
		default: isNote = false;
	}

	if (isNote) {
		//std::cerr << "Connected key " << label << ", octave " << octave << ", note " << note << std::endl;
		keys[octave*12 + note] = zone;
		return;
	}

}
void
YC20Processor::addVerticalSlider(const char* label, FAUSTFLOAT* zone, FAUSTFLOAT init, FAUSTFLOAT min, FAUSTFLOAT max, FAUSTFLOAT step)
{
	std::string name(label);

	Control *tmp = controlPerLabel[name];

	if (tmp == 0) {
		std::cerr << "ERROR: no Control for DSP label " << name << std::endl;
		return;
	}

	tmp->setZone(zone);
}

void
YC20Processor::addHorizontalSlider(const char* label, FAUSTFLOAT* zone, FAUSTFLOAT init, FAUSTFLOAT min, FAUSTFLOAT max, FAUSTFLOAT step)
{
	addVerticalSlider(label, zone, init, min, max, step);
}

YC20Processor::YC20Processor()
	: processor(0)
	, ui(0)
{
	controlPerLabel["pitch"]       = new Control(50, -0.35, 1.15);
	controlPerLabel["volume"]      = new Control(7);
	controlPerLabel["bass volume"] = new Control(51);
	
	controlPerLabel["realism"]     = new Control(52);
	controlPerLabel["depth"]       = new Control(12);
	controlPerLabel["speed"]       = new Control(13);

	controlPerLabel["16' b"]       = new Control(14);
	controlPerLabel["8' b"]        = new Control(15);
	controlPerLabel["bass manual"] = new Control(23);

	controlPerLabel["16' i"]       = new Control(2);
	controlPerLabel["8' i"]        = new Control(3);
	controlPerLabel["4' i"]        = new Control(4);
	controlPerLabel["2 2/3' i"]    = new Control(5);
	controlPerLabel["2' i"]        = new Control(6);
	controlPerLabel["1 3/5' i"]    = new Control(8);
	controlPerLabel["1' i"]        = new Control(9);

	controlPerLabel["balance"]     = new Control(16);
	controlPerLabel["bright"]      = new Control(17);

	controlPerLabel["16' ii"]      = new Control(18);
	controlPerLabel["8' ii"]       = new Control(19);
	controlPerLabel["4' ii"]       = new Control(20);
	controlPerLabel["2' ii"]       = new Control(21);

	controlPerLabel["percussive"]  = new Control(22);
}

YC20Processor::~YC20Processor()
{
	for (std::map<std::string, Control *>::iterator i = controlPerLabel.begin(); i !=  controlPerLabel.end(); ++i) {
		delete i->second;
	}

	if (ui != 0) {
		delete ui;
		ui = 0;
	}

	if (processor != 0) {
		deleteDSP(processor);
		processor = 0;
	}
}


void
YC20Processor::setDSP(dsp *dsp)
{
	processor = dsp;
	processor->buildUserInterface(this);
}

